### PR TITLE
EVG-19853: Log which Actors run in each Phase, at debug log level, in dry-run mode

### PR DESCRIPTION
**Pull request by PyGeek03**
**Jira Ticket:** EVG-19853

**Whats Changed:**  
Log which Actors run in each Phase, at debug log level, in Genny workload's dry-run mode.

For example, here's part of the log output for [TimeSeriesSort.yml](https://github.com/mongodb/genny/blob/f014c99eb7681e7b337b12e586469137e3a2343f/src/workloads/query/TimeSeriesSort.yml) (the full output can be found [here](https://privatebin.corp.mongodb.com/?0e99cb425102bff1#BDMZmq79DpUxhad8ZBZUezGCYjPkDAiMcLDWkDyKvT6J)):

```shell
> ./run-genny workload --dry-run -v debug src/workloads/query/TimeSeriesSort.yml 
...
[2023-07-04 15:50:37.341091] [0x00000001e8a9de00] [info]    Workload context constructed without errors.
[2023-07-04 15:50:37.342075] [0x00000001e8a9de00] [debug]   Phase 0 Actors:
[2023-07-04 15:50:37.342095] [0x00000001e8a9de00] [debug]   1) RunCommand.CreateTimeSeriesCollection
[2023-07-04 15:50:37.342102] [0x00000001e8a9de00] [debug]   Phase 1 Actors:
[2023-07-04 15:50:37.342108] [0x00000001e8a9de00] [debug]   1) RunCommand.CreateTimeSeriesCollection
[2023-07-04 15:50:37.342114] [0x00000001e8a9de00] [debug]   2) Loader.InsertData
[2023-07-04 15:50:37.342119] [0x00000001e8a9de00] [debug]   Phase 2 Actors:
[2023-07-04 15:50:37.342125] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342130] [0x00000001e8a9de00] [debug]   Phase 3 Actors:
[2023-07-04 15:50:37.342135] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
[2023-07-04 15:50:37.342140] [0x00000001e8a9de00] [debug]   Phase 4 Actors:
[2023-07-04 15:50:37.342145] [0x00000001e8a9de00] [debug]   1) RunCommand.IndexCollection
[2023-07-04 15:50:37.342150] [0x00000001e8a9de00] [debug]   Phase 5 Actors:
[2023-07-04 15:50:37.342155] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342160] [0x00000001e8a9de00] [debug]   Phase 6 Actors:
[2023-07-04 15:50:37.342165] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
[2023-07-04 15:50:37.342170] [0x00000001e8a9de00] [debug]   Phase 7 Actors:
[2023-07-04 15:50:37.342267] [0x00000001e8a9de00] [debug]   1) RunCommand.IndexCollection
[2023-07-04 15:50:37.342278] [0x00000001e8a9de00] [debug]   Phase 8 Actors:
[2023-07-04 15:50:37.342285] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342291] [0x00000001e8a9de00] [debug]   Phase 9 Actors:
[2023-07-04 15:50:37.342296] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
[2023-07-04 15:50:37.342301] [0x00000001e8a9de00] [debug]   Phase 10 Actors:
[2023-07-04 15:50:37.342307] [0x00000001e8a9de00] [debug]   1) RunCommand.IndexCollection
[2023-07-04 15:50:37.342312] [0x00000001e8a9de00] [debug]   Phase 11 Actors:
[2023-07-04 15:50:37.342317] [0x00000001e8a9de00] [debug]   1) Loader.InsertData
[2023-07-04 15:50:37.342322] [0x00000001e8a9de00] [debug]   Phase 12 Actors:
[2023-07-04 15:50:37.342327] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342332] [0x00000001e8a9de00] [debug]   Phase 13 Actors:
[2023-07-04 15:50:37.342408] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
[2023-07-04 15:50:37.342414] [0x00000001e8a9de00] [debug]   Phase 14 Actors:
[2023-07-04 15:50:37.342420] [0x00000001e8a9de00] [debug]   1) Loader.InsertData
[2023-07-04 15:50:37.342425] [0x00000001e8a9de00] [debug]   Phase 15 Actors:
[2023-07-04 15:50:37.342430] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342434] [0x00000001e8a9de00] [debug]   Phase 16 Actors:
[2023-07-04 15:50:37.342439] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
[2023-07-04 15:50:37.342445] [0x00000001e8a9de00] [debug]   Phase 17 Actors:
[2023-07-04 15:50:37.342450] [0x00000001e8a9de00] [debug]   1) RunCommand.IndexCollection
[2023-07-04 15:50:37.342455] [0x00000001e8a9de00] [debug]   2) Loader.InsertData
[2023-07-04 15:50:37.342460] [0x00000001e8a9de00] [debug]   Phase 20 Actors:
[2023-07-04 15:50:37.342465] [0x00000001e8a9de00] [debug]   1) QuiesceActor.Quiesce
[2023-07-04 15:50:37.342469] [0x00000001e8a9de00] [debug]   Phase 21 Actors:
[2023-07-04 15:50:37.342519] [0x00000001e8a9de00] [debug]   1) RunCommand.Queries
...
```

Doing phase logging in dry-run has 2 advantages: 
1) It avoids impacting normal Genny workloads on Evergreen
2) It is very quick to run locally.

In fact, it runs so quick, I find it quite enjoyable to run it with different workload YML files to have an idea of what each workload does.

A big difficulty with implementing this in dry-run is that, figuring out in which Phase an actor doesn't run (a `nop` Phase) requires the `PhaseContext` object, while figuring out which metrics an actor collect requires the `Actor` object, and there currently exists no link between `PhaseContext` and `Actor`. A potential solution is to add a member attribute to `Actor` that's a reference to this actor's `ActorContext` (which owns `PhaseContext`'s), however: 

1) For this ticket, my understanding is that the users just need to know which Actors run in each Phase. I've thought about logging the Actor's Operations, but the problem is that Operations is not part of standard Actor features, so only `CrudActor` and `RunCommand` support it while other Actors don't. I could use `PhaseContext`'s YML Node member attribute to check if there's an "Operations" or "Operation" node and print out its "OperationName" child node, but it feels brittle. 
2) For workloads with lots of Phases and many Actors in each Phase, logging out all the metrics collected by each Actor can be long and confusing to read. For example, [LLTMixed.yml](https://github.com/mongodb/genny/blob/4e6e1c7da4d830f44fbf4a65c4d4e3bae2011a6a/src/workloads/transactions/LLTMixed.yml) has 23 Phases, with 6 Actors running in 3 Phases (9, 15, 21)  and 5 Actors running in another 3 Phases (6, 12, 18). Here's part of this workload's log output (full output [here](https://privatebin.corp.mongodb.com/?70a6c9b8d62f6a49#EuFCk22Q3fGc9zZ364wFqCbf3ufnRKcauUX8pWaPSXFt)):

```shell
> ./run-genny workload --dry-run --verbosity debug src/workloads/transactions/LLTMixed.yml
...
[2023-07-04 14:28:10.463304] [0x00000001e8a9de00] [info]    Workload context constructed without errors.
[2023-07-04 14:28:10.466119] [0x00000001e8a9de00] [debug]   Phase 0 Actors:
[2023-07-04 14:28:10.466141] [0x00000001e8a9de00] [debug]   1) AdminCommand.SetTransactionLifetimeLimit
[2023-07-04 14:28:10.466149] [0x00000001e8a9de00] [debug]   2) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466155] [0x00000001e8a9de00] [debug]   Phase 1 Actors:
[2023-07-04 14:28:10.466160] [0x00000001e8a9de00] [debug]   1) Loader.InitialLoad
[2023-07-04 14:28:10.466166] [0x00000001e8a9de00] [debug]   2) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466171] [0x00000001e8a9de00] [debug]   Phase 2 Actors:
[2023-07-04 14:28:10.466176] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466181] [0x00000001e8a9de00] [debug]   Phase 3 Actors:
[2023-07-04 14:28:10.466186] [0x00000001e8a9de00] [debug]   1) Loader.SecondLoadAfterIndexes
[2023-07-04 14:28:10.466191] [0x00000001e8a9de00] [debug]   2) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466196] [0x00000001e8a9de00] [debug]   Phase 4 Actors:
[2023-07-04 14:28:10.466201] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466206] [0x00000001e8a9de00] [debug]   Phase 5 Actors:
[2023-07-04 14:28:10.466211] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466216] [0x00000001e8a9de00] [debug]   Phase 6 Actors:
[2023-07-04 14:28:10.466221] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466226] [0x00000001e8a9de00] [debug]   2) CrudActor.Short.Insert.Baseline
[2023-07-04 14:28:10.466231] [0x00000001e8a9de00] [debug]   3) CrudActor.Short.Query.Baseline
[2023-07-04 14:28:10.466335] [0x00000001e8a9de00] [debug]   4) CrudActor.Short.Update.Baseline
[2023-07-04 14:28:10.466345] [0x00000001e8a9de00] [debug]   5) CrudActor.Short.Remove.Baseline
[2023-07-04 14:28:10.466351] [0x00000001e8a9de00] [debug]   Phase 7 Actors:
[2023-07-04 14:28:10.466357] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466363] [0x00000001e8a9de00] [debug]   Phase 8 Actors:
[2023-07-04 14:28:10.466368] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466373] [0x00000001e8a9de00] [debug]   Phase 9 Actors:
[2023-07-04 14:28:10.466378] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466384] [0x00000001e8a9de00] [debug]   2) CrudActor.Short.Insert.Benchmark
[2023-07-04 14:28:10.466389] [0x00000001e8a9de00] [debug]   3) CrudActor.Short.Query.Benchmark
[2023-07-04 14:28:10.466394] [0x00000001e8a9de00] [debug]   4) CrudActor.Short.Update.Benchmark
[2023-07-04 14:28:10.466399] [0x00000001e8a9de00] [debug]   5) CrudActor.Short.Remove.Benchmark
[2023-07-04 14:28:10.466458] [0x00000001e8a9de00] [debug]   6) CollectionScanner.Short.Scan.Snapshot
[2023-07-04 14:28:10.466464] [0x00000001e8a9de00] [debug]   Phase 10 Actors:
[2023-07-04 14:28:10.466470] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466475] [0x00000001e8a9de00] [debug]   Phase 11 Actors:
[2023-07-04 14:28:10.466480] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466485] [0x00000001e8a9de00] [debug]   Phase 12 Actors:
[2023-07-04 14:28:10.466490] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466495] [0x00000001e8a9de00] [debug]   2) CrudActor.Medium.Insert.Baseline
[2023-07-04 14:28:10.466517] [0x00000001e8a9de00] [debug]   3) CrudActor.Medium.Query.Baseline
[2023-07-04 14:28:10.466530] [0x00000001e8a9de00] [debug]   4) CrudActor.Medium.Update.Baseline
[2023-07-04 14:28:10.466538] [0x00000001e8a9de00] [debug]   5) CrudActor.Medium.Remove.Baseline
[2023-07-04 14:28:10.466581] [0x00000001e8a9de00] [debug]   Phase 13 Actors:
[2023-07-04 14:28:10.466592] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466598] [0x00000001e8a9de00] [debug]   Phase 14 Actors:
[2023-07-04 14:28:10.466604] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466610] [0x00000001e8a9de00] [debug]   Phase 15 Actors:
[2023-07-04 14:28:10.466615] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466621] [0x00000001e8a9de00] [debug]   2) CrudActor.Medium.Insert.Benchmark
[2023-07-04 14:28:10.466626] [0x00000001e8a9de00] [debug]   3) CrudActor.Medium.Query.Benchmark
[2023-07-04 14:28:10.466632] [0x00000001e8a9de00] [debug]   4) CrudActor.Medium.Update.Benchmark
[2023-07-04 14:28:10.466637] [0x00000001e8a9de00] [debug]   5) CrudActor.Medium.Remove.Benchmark
[2023-07-04 14:28:10.466642] [0x00000001e8a9de00] [debug]   6) CollectionScanner.Medium.Scan.Snapshot
[2023-07-04 14:28:10.466684] [0x00000001e8a9de00] [debug]   Phase 16 Actors:
[2023-07-04 14:28:10.466694] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466701] [0x00000001e8a9de00] [debug]   Phase 17 Actors:
[2023-07-04 14:28:10.466707] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466712] [0x00000001e8a9de00] [debug]   Phase 18 Actors:
[2023-07-04 14:28:10.466718] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466723] [0x00000001e8a9de00] [debug]   2) CrudActor.Long.Insert.Baseline
[2023-07-04 14:28:10.466729] [0x00000001e8a9de00] [debug]   3) CrudActor.Long.Query.Baseline
[2023-07-04 14:28:10.466734] [0x00000001e8a9de00] [debug]   4) CrudActor.Long.Update.Baseline
[2023-07-04 14:28:10.466739] [0x00000001e8a9de00] [debug]   5) CrudActor.Long.Remove.Baseline
[2023-07-04 14:28:10.466744] [0x00000001e8a9de00] [debug]   Phase 19 Actors:
[2023-07-04 14:28:10.466750] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
[2023-07-04 14:28:10.466816] [0x00000001e8a9de00] [debug]   Phase 20 Actors:
[2023-07-04 14:28:10.466823] [0x00000001e8a9de00] [debug]   1) CollectionScanner.WarmUp
[2023-07-04 14:28:10.466828] [0x00000001e8a9de00] [debug]   Phase 21 Actors:
[2023-07-04 14:28:10.466844] [0x00000001e8a9de00] [debug]   1) LoggingActor.LoggingActor
[2023-07-04 14:28:10.466850] [0x00000001e8a9de00] [debug]   2) CrudActor.Long.Insert.Benchmark
[2023-07-04 14:28:10.466856] [0x00000001e8a9de00] [debug]   3) CrudActor.Long.Query.Benchmark
[2023-07-04 14:28:10.466862] [0x00000001e8a9de00] [debug]   4) CrudActor.Long.Update.Benchmark
[2023-07-04 14:28:10.466867] [0x00000001e8a9de00] [debug]   5) CrudActor.Long.Remove.Benchmark
[2023-07-04 14:28:10.466873] [0x00000001e8a9de00] [debug]   6) CollectionScanner.Long.Scan.Snapshot
[2023-07-04 14:28:10.466878] [0x00000001e8a9de00] [debug]   Phase 22 Actors:
[2023-07-04 14:28:10.466884] [0x00000001e8a9de00] [debug]   1) QuiesceActor.QuiescePhase
...
```

Although on Evergreen, a genny patch's dry-run task is its longest task and hence dominates the patch runtime, the impact of this PR is negligible, because phase logging is done at DEBUG level while the patch runs Genny at INFO log level. This is proven by the Genny patch for this PR, as can be seen in this [dry-run task log](https://parsley.mongodb.com/evergreen/genny_amazon2_arm64_t_compile_and_dry_run_asan_patch_a3bebe0144abbc04e42d690644ec31d8fa59729e_64a3a07c3e8e864f1fd67815_23_07_04_04_30_54/0/task?bookmarks=0,12737).

**Patch testing results:**  
- [sys-perf, big_update_10k workload dry-run](https://spruce.mongodb.com/task/sys_perf_linux_3_node_replSet.2022_11_big_update_10k_patch_b7fecb4445d3e54a601e4e4dd79b9d57d082117f_649b8e1ba4cf47b96356eed5_23_06_28_02_52_44/logs?execution=0). For this patch, DSI is modified to run in dry-run, hence the failure at DSI analysis step. The Phases and Actors in each Phase start being logged [here](https://parsley.mongodb.com/evergreen/sys_perf_linux_3_node_replSet.2022_11_big_update_10k_patch_b7fecb4445d3e54a601e4e4dd79b9d57d082117f_649b8e1ba4cf47b96356eed5_23_06_28_02_52_44/1/task?bookmarks=0,6151&shareLine=5938). 
- [sys-perf, big_update_10k workload normal run](https://spruce.mongodb.com/task/sys_perf_linux_3_node_replSet.2022_11_big_update_10k_patch_b7fecb4445d3e54a601e4e4dd79b9d57d082117f_64a3c6250305b91723e2af84_23_07_04_07_12_53/logs?execution=0).

**Related PRs:**   
Previous PR that does phase logging in Genny workload's normal mode: #935

### CODE REVIEW

**Comment by PyGeek03** [comment]
I've used Perf Analyzer to test whether this PR introduces any performance regression, and the result is no:

https://performance-analyzer.server-tig.prod.corp.mongodb.com/perf-analyzer-viz/?comparison_id=110f5a84-3d75-4483-81c5-c818f21c0afd&percent_filter=0%7C%7C100&z_filter=0%7C%7C10

File: src/gennylib/include/gennylib/context.hpp [hunk 1]
```diff
@@ -158,6 +158,9 @@ class ExternalPhaseCoordinator {
     const int m_socket;
 };
 
+// The compiler is not happy if we don't forward-declare it here; see impl below
```
Comment (thessem) [inline src/gennylib/include/gennylib/context.hpp:161]: Noting that this comment and "strange-ness" has not been added in this PR, it's just been moved up

**Review by thessem — APPROVED** [review]
Thank you for the thorough investigation on this. Appreciated. I think we found the right solution for the problem in the end.

I would like to give @rtimmons a chance to take a look at this if he likes, given he came up with the original suggestion.

**Comment by PyGeek03** [comment]
Hi @mattBoros @HanaPearlman, could you take a look at this PR to see if this satisfy your requirements for EVG-19853? Thank you!

**Review by mattBoros — APPROVED** [review]
LGTM! This will be very helpful in developing Genny benchmarks, I appreciate all the work you've done!

**Comment by PyGeek03** [comment]
evergreen merge

**Pull request merged**